x/programs/cmd/simulator/src/lib.rs

```diff
@@ -233,46 +206,47 @@ impl Client {
             panic!("Simulator binary not found, must rebuild simulator");
         }
 
-        Self { path }
-    }
-
-    /// Runs a [Plan] against the simulator and returns vec of result.
-    /// # Errors
-    ///
-    /// Returns an error if the serialization or plan fails.
-    pub fn run_plan(self, plan: Plan) -> Result<Vec<PlanResponse>, ClientError> {
-        let Child {
-            mut stdin,
-            mut stdout,
-            ..
-        } = Command::new(self.path)
+        let Child { stdin, stdout, .. } = Command::new(path)
             .arg("interpreter")
             .arg("--cleanup")
             .arg("--log-level")
             .arg("error")
             .stdin(Stdio::piped())
             .stdout(Stdio::piped())
-            .spawn()?;
+            .spawn()
+            .expect("failed to run the simulator in interpreter mode");
 
-        let writer = stdin.as_mut().ok_or(ClientError::StdIo)?;
-        let reader = stdout.as_mut().ok_or(ClientError::StdIo)?;
+        let writer = stdin.expect("stdin has been captured");
+        let reader = stdout.expect("stdout has been captured");
 
         let responses = BufReader::new(reader)
             .lines()
             .map(|line| serde_json::from_str(&line?).map_err(Into::into));
 
-        let mut child = SimulatorChild { writer, responses };
+        let child = SimulatorChild { writer, responses };
+
+        Client { child }
```

Comment:
This shouldn't be nested like this, the `Client` doesn't provide any functionality, it only delegates to the `SimulatorChild`. In other words, this type is more complex than it needs to be. What I would do is have a struct called a `ClientBuilder` with a `ClientBuilder::new` that takes in a path and a `ClientBuilder::try_build` method that returns a `Client<impl Write, impl Iterator<Item = Result<PlanResponse, ClientError>>>` (which is a rename on the `SimulatorChild`). 

That would fix the type issues you're having above.

Reply:
Thanks a lot for this suggestion, I did all of your suggested stuff. The `SimulatorChild` has been flattened inside the `Client` structure and it's now much easier.